### PR TITLE
Adds SDQL2 and a backend VV refactor

### PR DESCRIPTION
**Pull request by Zonespace27**
## About The Pull Request
This PR does two things:

- Lovingly ports SDQL, the Structured Datum Query Language.
- Refactors a lot of the back end of VV since SDQL required it

## What is SDQL?

SDQL is a SQL-styled method of acting on large (or small) groups of datums, allowing for massively granular control over mass data collection, VVing, and proc calling.

For example, the following SDQL query is as basic as they get, but very effective:
`SELECT /obj/item/weapon/gun`
Which gets a list of every gun in the world and returns it in a list.
image

Next,
`SELECT /obj/item/weapon/gun WHERE src.x == 101`
Which gets a list of all guns that are at x 101, regardless of y or z.
image

And finally,
`SELECT /obj/item/weapon/gun WHERE global._istype(src.loc, /turf/open/floor)`
Which will get a list of all guns whose loc is a floor tile.
image

But this is just SELECT, you can do fun stuff like:
`CALL gib() ON /mob/living/carbon/human`
Which'll gib everyone at once, fun.

For more details, please see:

- The extensive documentation in SDQL2.dm
- [This document I wrote for /tg/ codebases, that still mostly applies here.](https://hackmd.io/5yrXqD44SYqpzJULxXh5Rg)

## The refactor part

- Cut out a lot of repeat code from VV
- Preventing varedits is now per-type with `vv_edit_var()`

## Why It's Good For The Game
It's more admin tools (especially a very versatile one), and refactors are good

## Changelog

:cl:
admin: Added SDQL2 tools, allowing for more granular mass VV/proc calling/data collecting.
refactor: Refactored VV backend
/:cl:

### CODE REVIEW

File: code/modules/admin/verbs/SDQL2/SDQL_2.dm [hunk 1]
```diff
@@ -0,0 +1,1234 @@
...
+	return ..()
+
+/datum/sdql2_query/proc/get_query_text()
+	var/list/out = list()
+	recursive_list_print(out, query_tree)
+	return out.Join()
+
+/proc/recursive_list_print(list/output = list(), list/input, datum/callback/datum_handler, datum/callback/atom_handler)
+	output += "\[ "
+	for(var/i in 1 to lenght(input))
```
Comment (harryob) [inline code/modules/admin/verbs/SDQL2/SDQL_2.dm:372]: linter fail boop

File: code/modules/admin/verbs/SDQL2/SDQL_2.dm [hunk 2]
```diff
@@ -0,0 +1,1234 @@
...
+			var/turf/T = A.loc
+			var/area/gotten_area
+			if(istype(T))
+				text_list += " <font color='gray'>at</font> [T] [ADMIN_COORDJMP(T)]"
+				gotten_area = T.loc
+			else
+				var/turf/final = get_turf(T) //Recursive, hopefully?
+				if(istype(final))
+					text_list += " <font color='gray'>at</font> [final] [ADMIN_COORDJMP(final)]"
+					a = final.loc
```
Comment (harryob) [inline code/modules/admin/verbs/SDQL2/SDQL_2.dm:738]:
```suggestion
					gotten_area = final.loc
```

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/modules/admin/verbs/SDQL2/SDQL_2.dm (hunk 1 above)
Comment (Zonespace27) [inline code/modules/admin/verbs/SDQL2/SDQL_2.dm:372]:
```suggestion
	for(var/i in 1 to length(input))
```

Comment (Zonespace27) [reply code/modules/admin/verbs/SDQL2/SDQL_2.dm:372]: ty

**Review by harryob — APPROVED** [review]
😬 i’m probably still missing something but… i’d like a smidge of TM before merging for sanities sake

File: code/modules/admin/callproc.dm [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+		return call("/proc/[procname]")(arglist(arguments))
+	else if(target != world)
+		return call(target, procname)(arglist(arguments))
+	else
+		log_admin("[key_name(usr)] attempted to call world/proc/[procname] with arguments: [english_list(arguments)]")
+
+/proc/IsAdminAdvancedProcCall()
+	return (GLOB.AdminProcCaller && GLOB.AdminProcCaller == usr?.client?.ckey) || (GLOB.AdminProcCallHandler && usr == GLOB.AdminProcCallHandler)
+
+/client/proc/callproc_datum(datum/called_datum)
```
Comment (Segrain) [inline code/modules/admin/callproc.dm:226]:
```suggestion
/client/proc/callproc_datum(datum/called_datum as null|area|mob|obj|turf)
```
For whatever reason, Datum ProcCall does not appear on right click menu the way Atom ProcCall did. I know nothing about underlying Byond sorceries here, but any chance this difference between definitions might be responsible?

**Comment by Podrick-Equus** [comment]
This PR seems to remove some functionality. I can no longer mass edit certain variables to certain values through the VV menu, such as mass editing "Hull = 1" to "Hull = 0"

**Comment by Zonespace27** [comment]
> This PR seems to remove some functionality. I can no longer mass edit certain variables to certain values through the VV menu, such as mass editing "Hull = 1" to "Hull = 0"

Unintended (callproc datum too), will look into later today

**Comment by Segrain** [comment]
Github does not allow suggestions several lines away from code being modified by PR, so left one more in form of the commit above instead (which Github apparently allows allowing to do). Should not break anything.

**Comment by Zonespace27** [comment]
> This PR seems to remove some functionality. I can no longer mass edit certain variables to certain values through the VV menu, such as mass editing "Hull = 1" to "Hull = 0"

This bug has been fixed, queries have not yet

File: code/modules/admin/verbs/SDQL2/SDQL_2.dm [hunk 4]
```diff
@@ -0,0 +1,1234 @@
...
+			log_admin(msg)
+			query.ARun()
+
+	var/finished = FALSE
+	var/objs_all = 0
+	var/objs_eligible = 0
+	var/selectors_used = FALSE
+	var/list/combined_refs = list()
+	do
+		CHECK_TICK
```
[the remark this answers is not in the document]
Comment (fira) [reply code/modules/admin/verbs/SDQL2/SDQL_2.dm:266]: Oh yeah that's a good point
Can make this a `stoplag(world.tick_lag)`
This really looks like something that should be controller-backed

**Review by harryob — CHANGES_REQUESTED** [review]
per silicons, i have faith in their understanding of sdql here

File: code/modules/admin/verbs/SDQL2/SDQL_2.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (Zonespace27) [reply code/modules/admin/verbs/SDQL2/SDQL_2.dm:266]: resolved

File: code/modules/admin/callproc.dm [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+		return call("/proc/[procname]")(arglist(arguments))
+	else if(target != world)
+		return call(target, procname)(arglist(arguments))
+	else
+		log_admin("[key_name(usr)] attempted to call world/proc/[procname] with arguments: [english_list(arguments)]")
+
+/proc/IsAdminAdvancedProcCall()
+	return (GLOB.AdminProcCaller && GLOB.AdminProcCaller == usr?.client?.ckey) || (GLOB.AdminProcCallHandler && usr == GLOB.AdminProcCallHandler)
+
+/client/proc/callproc_datum(datum/called_datum as null|area|mob|obj|turf|datum)
```
Comment (harryob) [inline code/modules/admin/callproc.dm:227]: ci fail here btw

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/modules/admin/callproc.dm (hunk 5 above)
Comment (Zonespace27) [inline code/modules/admin/callproc.dm:227]:
```suggestion
/client/proc/callproc_datum(datum/called_datum as null|area|mob|obj|turf)
```

Comment (Zonespace27) [reply code/modules/admin/callproc.dm:227]: ty

**Review by harryob — APPROVED** [review]

**Comment by harryob** [comment]
fuck it

**Pull request merged**